azure/services/vmextensions/client.go

```diff
@@ -18,138 +18,103 @@ package vmextensions
-	azure.SetAutoRestClientDefaults(&vmextensionsClient.Client, authorizer)
-	return vmextensionsClient
+// newClient creates a new vm extensions client from an authorizer.
+func newClient(auth azure.Authorizer) (*azureClient, error) {
+	opts, err := azure.ARMClientOptions(auth.CloudEnvironment())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create virtualmachineextensions client options")
+	}
+	factory, err := armcompute.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create armcompute client factory")
+	}
+	return &azureClient{factory.NewVirtualMachineExtensionsClient()}, nil
 }
 
 // Get the specified virtual machine extension.
 func (ac *azureClient) Get(ctx context.Context, spec azure.ResourceSpecGetter) (result interface{}, err error) {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "vmextensions.AzureClient.Get")
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "vmextensions.azureClient.Get")
 	defer done()
 
-	return ac.vmextensions.Get(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName(), "")
+	resp, err := ac.vmextensions.Get(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName(), nil)
+	if err != nil {
+		return nil, err
+	}
+	return resp.VirtualMachineExtension, nil
 }
 
 // CreateOrUpdateAsync creates or updates a VM extension asynchronously.
-// It sends a PUT request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// It sends a PUT request to Azure and if accepted without error, the func will return a Poller which can be used to track the ongoing
 // progress of the operation.
-func (ac *azureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{}) (result interface{}, future azureautorest.FutureAPI, err error) {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "vmextensions.AzureClient.CreateOrUpdateAsync")
+func (ac *azureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, resumeToken string, parameters interface{}) (result interface{}, poller *runtime.Poller[armcompute.VirtualMachineExtensionsClientCreateOrUpdateResponse], err error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "vmextensions.azureClient.CreateOrUpdateAsync")
 	defer done()
 
-	vmextension, ok := parameters.(compute.VirtualMachineExtension)
+	vmextension, ok := parameters.(armcompute.VirtualMachineExtension)
 	if !ok {
-		return nil, nil, errors.Errorf("%T is not a compute.VirtualMachineExtension", parameters)
+		return nil, nil, errors.Errorf("%T is not an armcompute.VirtualMachineExtension", parameters)
 	}
 
-	createFuture, err := ac.vmextensions.CreateOrUpdate(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName(), vmextension)
+	opts := &armcompute.VirtualMachineExtensionsClientBeginCreateOrUpdateOptions{ResumeToken: resumeToken}
```

Comment:
Another consistency nit: there is a log message in a previously merged PR https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3858/files#diff-0e0342ee34be437e8b2443243544f4d3a79fc4655bf9552bffba2f187f4d18ebR83

Reply:
That log message was meant more for debugging (I realized after it merged), so I've kept it out of subsequent SDKv2 PRs like this one. I can remove it from natgateways too at some point.